### ISSUE TITLE
Unable to change the mysql root password

### ISSUE THREAD

**Issue by ltarasiewicz** [infOpen/ansible-role-mysql-server#2]
Hello,

Thanks for this great, easy to use ansible role.

However, I'm not sure what the purpose of `mysqlserver_root_db_old_password` is. I attempted to change the mysql root password that I set when I run the role on the server for the first time. 

Once the role run, changing the password would not go through, even with the abovementioned variable populated correctly. It'd error out with:

> msg: unable to connect to database, check login_user and login_password are correct or ~/.my.cnf has the credentials

I guess that changing the mysql root password is not so straightforward, but what'd then be the role of `mysqlserver_root_db_old_password`, if it's not for changing it?

**Comment by achaussier** [comment]
Hello,

Thanks for use it and your report. I just test with a vagrant box, and it's work for me, but I should add documentation for the root password.

When you change the root password, you set the actual password to "mysqlserver_root_db_old_password" and the future to "mysqlserver_root_db_password".
But, once the role with the password change ended, you could set "mysqlserver_root_db_old_password" with the new password also, else the role will try to change another time the root password, and will break because password has changed.

```
vagrant@vagrant-ubuntu-trusty-64:~$ cat test-mysql.yml 
---

- hosts : all
  vars :
    mysqlserver_root_db_old_password : ""
    mysqlserver_root_db_password     : "foo"
  roles:
    - achaussier.mysql-server

vagrant@vagrant-ubuntu-trusty-64:~$ ansible-playbook -i inventory test-mysql.yml --connection=local --sudo

PLAY [all] ******************************************************************** 

GATHERING FACTS *************************************************************** 
ok: [localhost]

TASK: [achaussier.mysql-server | Load the OS specific varibles] *************** 
ok: [localhost]

TASK: [achaussier.mysql-server | install mysql-server packages] *************** 
changed: [localhost] => (item=mysql-client,mysql-common,mysql-server,mysql-server-core,python-mysqldb)

TASK: [achaussier.mysql-server | stat datadir exists] ************************* 
ok: [localhost]

TASK: [achaussier.mysql-server | create binary logs directory] **************** 
changed: [localhost]

TASK: [achaussier.mysql-server | stop mysql if datadir move is needed] ******** 
skipping: [localhost]

TASK: [achaussier.mysql-server | get current datadir configuration] *********** 
skipping: [localhost]

TASK: [achaussier.mysql-server | move mysql data to data disk] **************** 
skipping: [localhost]

TASK: [achaussier.mysql-server | generate custom role config file] ************ 
changed: [localhost] => (item=/etc/ansible/roles/achaussier.mysql-server/templates/my.cnf.j2)

TASK: [achaussier.mysql-server | generate logrotate configuration] ************ 
changed: [localhost] => (item=/etc/ansible/roles/achaussier.mysql-server/templates/logrotate_mysql.j2)

TASK: [achaussier.mysql-server | enable mysql service on boot] **************** 
ok: [localhost]

TASK: [achaussier.mysql-server | update mysql root password for all localhost root accounts] *** 
changed: [localhost] => (item=127.0.0.1)
changed: [localhost] => (item=::1)
changed: [localhost] => (item=localhost)

TASK: [achaussier.mysql-server | update mysql root password for hostname root account] *** 
changed: [localhost]

TASK: [achaussier.mysql-server | manage mysql global accounts] **************** 
skipping: [localhost]

TASK: [achaussier.mysql-server | generate credentials files for global accounts] *** 
skipping: [localhost]

TASK: [achaussier.mysql-server | generate credentials files for additionnal accounts] *** 
skipping: [localhost]

NOTIFIED: [achaussier.mysql-server | restart mysql] *************************** 
changed: [localhost]

PLAY RECAP ******************************************************************** 
localhost                  : ok=11   changed=7    unreachable=0    failed=0   

vagrant@vagrant-ubuntu-trusty-64:~$ mysql -u root
ERROR 1045 (28000): Access denied for user 'root'@'localhost' (using password: NO)

vagrant@vagrant-ubuntu-trusty-64:~$ mysql -u root -pfoo
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 38
Server version: 5.5.46-0ubuntu0.14.04.2-log (Ubuntu)

Copyright (c) 2000, 2015, Oracle and/or its affiliates. All rights reserved.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> exit

# Change password

vagrant@vagrant-ubuntu-trusty-64:~$ cat test-mysql.yml 
---

- hosts : all
  vars :
    mysqlserver_root_db_old_password : "foo"
    mysqlserver_root_db_password     : "bar"
  roles:
    - achaussier.mysql-server
vagrant@vagrant-ubuntu-trusty-64:~$ ansible-playbook -i inventory test-mysql.yml --connection=local --sudo

PLAY [all] ******************************************************************** 

GATHERING FACTS *************************************************************** 
ok: [localhost]

TASK: [achaussier.mysql-server | Load the OS specific varibles] *************** 
ok: [localhost]

TASK: [achaussier.mysql-server | install mysql-server packages] *************** 
changed: [localhost] => (item=mysql-client,mysql-common,mysql-server,mysql-server-core,python-mysqldb)

TASK: [achaussier.mysql-server | stat datadir exists] ************************* 
ok: [localhost]

TASK: [achaussier.mysql-server | create binary logs directory] **************** 
ok: [localhost]

TASK: [achaussier.mysql-server | stop mysql if datadir move is needed] ******** 
skipping: [localhost]

TASK: [achaussier.mysql-server | get current datadir configuration] *********** 
skipping: [localhost]

TASK: [achaussier.mysql-server | move mysql data to data disk] **************** 
skipping: [localhost]

TASK: [achaussier.mysql-server | generate custom role config file] ************ 
ok: [localhost] => (item=/etc/ansible/roles/achaussier.mysql-server/templates/my.cnf.j2)

TASK: [achaussier.mysql-server | generate logrotate configuration] ************ 
ok: [localhost] => (item=/etc/ansible/roles/achaussier.mysql-server/templates/logrotate_mysql.j2)

TASK: [achaussier.mysql-server | enable mysql service on boot] **************** 
ok: [localhost]

TASK: [achaussier.mysql-server | update mysql root password for all localhost root accounts] *** 
changed: [localhost] => (item=127.0.0.1)
changed: [localhost] => (item=::1)
changed: [localhost] => (item=localhost)

TASK: [achaussier.mysql-server | update mysql root password for hostname root account] *** 
changed: [localhost]

TASK: [achaussier.mysql-server | manage mysql global accounts] **************** 
skipping: [localhost]

TASK: [achaussier.mysql-server | generate credentials files for global accounts] *** 
skipping: [localhost]

TASK: [achaussier.mysql-server | generate credentials files for additionnal accounts] *** 
skipping: [localhost]

NOTIFIED: [achaussier.mysql-server | restart mysql] *************************** 
changed: [localhost]

PLAY RECAP ******************************************************************** 
localhost                  : ok=11   changed=4    unreachable=0    failed=0   

vagrant@vagrant-ubuntu-trusty-64:~$ mysql -u root -pfoo
ERROR 1045 (28000): Access denied for user 'root'@'localhost' (using password: YES)
vagrant@vagrant-ubuntu-trusty-64:~$ mysql -u root -pbar
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 37
Server version: 5.5.46-0ubuntu0.14.04.2-log (Ubuntu)

Copyright (c) 2000, 2015, Oracle and/or its affiliates. All rights reserved.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> ^CCtrl-C -- exit!
Aborted

# Now, set the same password and create a fake user to test :
vagrant@vagrant-ubuntu-trusty-64:~$ cat test-mysql.yml 
---

- hosts : all
  vars :
    mysqlserver_root_db_old_password : "bar"
    mysqlserver_root_db_password     : "bar"
    mysqlserver_global_accounts :
      - name       : monitoring
        password   : foobar
        host       : localhost
        state      : present
        files      :
          - path  : "/root/.my_monitoring.cnf"
            owner : root
            group : root
            mode  : 0400
        privileges :
          - "*.*:USAGE"
  roles:
    - achaussier.mysql-server

vagrant@vagrant-ubuntu-trusty-64:~$ ansible-playbook -i inventory test-mysql.yml --connection=local --sudo

PLAY [all] ******************************************************************** 

GATHERING FACTS *************************************************************** 
ok: [localhost]

TASK: [achaussier.mysql-server | Load the OS specific varibles] *************** 
ok: [localhost]

TASK: [achaussier.mysql-server | install mysql-server packages] *************** 
changed: [localhost] => (item=mysql-client,mysql-common,mysql-server,mysql-server-core,python-mysqldb)

TASK: [achaussier.mysql-server | stat datadir exists] ************************* 
ok: [localhost]

TASK: [achaussier.mysql-server | create binary logs directory] **************** 
ok: [localhost]

TASK: [achaussier.mysql-server | stop mysql if datadir move is needed] ******** 
skipping: [localhost]

TASK: [achaussier.mysql-server | get current datadir configuration] *********** 
skipping: [localhost]

TASK: [achaussier.mysql-server | move mysql data to data disk] **************** 
skipping: [localhost]

TASK: [achaussier.mysql-server | generate custom role config file] ************ 
ok: [localhost] => (item=/etc/ansible/roles/achaussier.mysql-server/templates/my.cnf.j2)

TASK: [achaussier.mysql-server | generate logrotate configuration] ************ 
ok: [localhost] => (item=/etc/ansible/roles/achaussier.mysql-server/templates/logrotate_mysql.j2)

TASK: [achaussier.mysql-server | enable mysql service on boot] **************** 
ok: [localhost]

TASK: [achaussier.mysql-server | update mysql root password for all localhost root accounts] *** 
ok: [localhost] => (item=127.0.0.1)
ok: [localhost] => (item=::1)
ok: [localhost] => (item=localhost)

TASK: [achaussier.mysql-server | update mysql root password for hostname root account] *** 
ok: [localhost]

TASK: [achaussier.mysql-server | manage mysql global accounts] **************** 
changed: [localhost] => (item={'files': [{'owner': 'root', 'path': '/root/.my_monitoring.cnf', 'group': 'root', 'mode': 256}], 'state': 'present', 'name': 'monitoring', 'privileges': ['*.*:USAGE'], 'password': 'foobar', 'host': 'localhost'})

TASK: [achaussier.mysql-server | generate credentials files for global accounts] *** 
changed: [localhost] => (item=({'name': 'monitoring', 'privileges': ['*.*:USAGE'], 'state': 'present', 'host': 'localhost', 'password': 'foobar'}, {'owner': 'root', 'path': '/root/.my_monitoring.cnf', 'group': 'root', 'mode': 256}))

TASK: [achaussier.mysql-server | generate credentials files for additionnal accounts] *** 
skipping: [localhost]

NOTIFIED: [achaussier.mysql-server | restart mysql] *************************** 
changed: [localhost]

PLAY RECAP ******************************************************************** 
localhost                  : ok=13   changed=4    unreachable=0    failed=0   

vagrant@vagrant-ubuntu-trusty-64:~$ mysql -u monitoring -pfoobar
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 36
Server version: 5.5.46-0ubuntu0.14.04.2-log (Ubuntu)

Copyright (c) 2000, 2015, Oracle and/or its affiliates. All rights reserved.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> ^CCtrl-C -- exit!
Aborted
vagrant@vagrant-ubuntu-trusty-64:~$
```

**Comment by ltarasiewicz** [comment]
All is good, indeed. I've been overriding `mysqlserver_root_db_password` in my `group_vars` not even knowing it. This was the cause of my problem. Otherwise it's all good. However, the funny thing is that I'm not able to log in with `mysql -u foor -pbar`, but only with `mysql -u foor -p` , and then providing the pass on the separate line. That is weird:)

**Comment by achaussier** [comment]
Funny :)

I'll add a "testing" task into playbook to check if the old password works. With that, the root password change tasks will run only if the "testing" task not fails. The role will have better idempotence.

I've add Vagrant management to do local test and dev.

A new release with all these changes will come today :)

**Comment by achaussier** [comment]
I can close this issue, have a nice day.

**Comment by achaussier** [comment]
I've released 0.2.0 if you want.
